### PR TITLE
Enable Merging of BibDatabases

### PR DESCRIPTION
**Pull request by DominikVoigt**
This PR adds a method to the BibDatabases that allows instances to be merged with other instances.
This merging will not introduce duplicates if an entry is contained in both databases.

- [ ] Change in CHANGELOG.md described (if applicable)
- [x] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [ ] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, submitted a pull request to the documentation repository.

### CODE REVIEW

**Comment by koppor** [comment]
The linked code refs https://github.com/JabRef/jabref/issues/6488

Currently, I understand the new code better than the looong linked code.

We also have `BibDatabaseDiff`, which doesn't seem to be referenced in any code. (Example use: https://github.com/koppor/jabref/pull/442/files#diff-58f2d74f80b59a9ba6468243b879c428R68)

**Comment by koppor** [comment]
We merged the merge methods at the cose of "some" undo/redo (at the whole import 😟)

undo/redo refs https://github.com/koppor/jabref/issues/453

**Review by koppor — APPROVED** [review]
Commit issues resolved. LGTM.

File: src/main/java/org/jabref/gui/importer/ImportAction.java [hunk 1]
```diff
@@ -135,55 +130,14 @@ private ParserResult mergeImportResults(List<ImportFormatReader.UnknownFormatImp
                 continue;
             }
             ParserResult parserResult = importResult.parserResult;
-            List<BibEntry> entries = parserResult.getDatabase().getEntries();
-            resultDatabase.insertEntries(entries);
+            new BibDatabaseMerger().merge(resultDatabase, parserResult.getDatabase());
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/importer/ImportAction.java:133]: The merge method below checks for duplicates etc, but this is already done in the import dialog where the user can also decide to overwrite the duplication check. This is behavior that we should keep, i.e. here we really need to add the entries from one db into the other one.

File: src/main/java/org/jabref/logic/bibtex/BibDatabaseMerger.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
+package org.jabref.logic.bibtex;
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/logic/bibtex/BibDatabaseMerger.java:1]: Since this is not related to writing bibtex (as the rest of the package), I suggest to move it together with the duplication check to a new `logic.database` package in parallel to the existing `model.database`. @JabRef/developers opinions?

File: src/main/java/org/jabref/logic/bibtex/BibDatabaseMerger.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+import org.jabref.model.database.BibDatabase;
+import org.jabref.model.database.BibDatabaseModeDetection;
+import org.jabref.model.entry.BibEntry;
+import org.jabref.model.entry.BibEntryTypesManager;
+import org.jabref.model.entry.BibtexString;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class BibDatabaseMerger {
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/logic/bibtex/BibDatabaseMerger.java:15]: `DatabaseMerger` should be sufficient...I have actually no idea why we added the prefix bib in some instances...

File: src/test/java/org/jabref/logic/bibtex/BibDatabaseMergerTest.java [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+class BibDatabaseMergerTest {
+
+    @Test
+    void mergeAddsNonDuplicateEntries() {
+        // Entries 2 and 3 are identical
+        BibEntry entry1 = new BibEntry()
+                .withField(StandardField.AUTHOR, "Stephen Blaha")
```
Comment (tobiasdiez) [inline src/test/java/org/jabref/logic/bibtex/BibDatabaseMergerTest.java:22]: Please reduce these examples to a minimum. I don't think you need to have 4 entries with full information

File: src/main/java/org/jabref/model/metadata/MetaData.java [hunk 5]
```diff
@@ -245,6 +254,54 @@ public void clearSaveOrderConfig() {
         postChange();
     }
 
+    /**
+     * @param other           the metaData to merge into this one
+     * @param otherFilename   the filename of the other library. Pass "unknown" if not known.
+     * @param allOtherEntries list of all other entries
+     */
+    public void merge(MetaData other, String otherFilename, List<BibEntry> allOtherEntries) {
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/model/metadata/MetaData.java:262]: Please move these merge metadata methods to the database merger class. There you can also add a merge method operating on `BibDatabaseContext` objects, to have an easy way to merge two databases including all their metadata.

File: src/test/java/org/jabref/logic/bibtex/BibDatabaseMergerTest.java [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+        new BibDatabaseMerger().mergeStrings(target, source);
+        // Use string representation to compare since the id will not match
+        List<String> resultStringsSorted = target.getStringValues()
+                                                 .stream()
+                                                 .map(BibtexString::toString)
+                                                 .sorted()
+                                                 .collect(Collectors.toList());
+
+        assertEquals(List.of(targetString1.toString(), targetString2.toString()), resultStringsSorted);
+    }
```
Comment (tobiasdiez) [inline src/test/java/org/jabref/logic/bibtex/BibDatabaseMergerTest.java:133]: Please also include tests for the merged groups.

**Review by tobiasdiez — CHANGES_REQUESTED** [review]
Thanks. The code looks mostly good to me. I've a few comments and suggestions for improvement.

File: src/main/java/org/jabref/logic/bibtex/BibDatabaseMerger.java (hunk 2 above)
Comment (DominikVoigt) [reply src/main/java/org/jabref/logic/bibtex/BibDatabaseMerger.java]: I moved it to the recommended package :).

File: src/test/java/org/jabref/logic/bibtex/BibDatabaseMergerTest.java (hunk 4 above)
Comment (DominikVoigt) [reply src/test/java/org/jabref/logic/bibtex/BibDatabaseMergerTest.java]: I reduced the number of entries and their information content.

**Comment by DominikVoigt** [comment]
Thanks for your comments :). I implemented the requested changes!

**Review by tobiasdiez — APPROVED** [review]
